File: programmers/kakao/level3/사라지는_발판.py
moves = ((0, 1), (-1, 0), (1, 0), (0, -1))

def game(board, aloc, bloc):
    game_results = []

    if board[bloc[0]][bloc[1]] == 0:
        return True, 0
    if board[aloc[0]][aloc[1]] == 0:
        return False, 0

    for move in moves:
        next_aloc = (aloc[0]+move[0], aloc[1]+move[1])
        try:
            if next_aloc[0] < 0 or next_aloc[1] < 0:
                raise IndexError
            next_board = board[next_aloc[0]][next_aloc[1]]
        except IndexError:
            continue

        if next_board == 1:
            board[aloc[0]][aloc[1]] = 0
            win, game_length = game(board, bloc, next_aloc)
            board[aloc[0]][aloc[1]] = 1
            game_results.append((not win, game_length+1))

    if len(game_results) == 0:
        return False, 0
    elif any(r[0] for r in game_results):
        return True, min(r[1] for r in game_results if r[0])
    else:
        return False, max(r[1] for r in game_results)

def solution(board, aloc, bloc):
    return game(board, aloc, bloc)[1]


board = [[1, 1, 1], [1, 1, 1], [1, 1, 1]]
aloc, bloc = [1, 0], [1, 2]
print(solution(board, aloc, bloc)) # 5

board = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
aloc, bloc = [1, 0], [1, 2]
print(solution(board, aloc, bloc)) # 4

board = [[1, 1, 1, 1, 1]]
aloc, bloc = [0, 0], [0, 4]
print(solution(board, aloc, bloc)) # 4

board = [[1]]
aloc, bloc = [0, 0], [0, 0]
print(solution(board, aloc, bloc)) # 0

File: Java/programmers/kakao/level3/사라지는발판.java
package CodeTest.Java.programmers.kakao.level3;

import java.util.*;

class Game {
    boolean win;
    int turn;

    Game(boolean win, int turn) {
        this.win = win;
        this.turn = turn;
    }
}

public class 사라지는발판 {
    private static int[][] moves = {{0,1},{-1,0},{1,0},{0,-1}};

    public int solution(int[][] board, int[] aloc, int[] bloc) {
        return simulation(board, aloc, bloc).turn;
    }

    private Game simulation(int[][] board, int[] aloc, int[] bloc) {
        List<Game> results = new ArrayList<>();

        if (board[bloc[0]][bloc[1]] == 0) return new Game(true, 0);
        if (board[aloc[0]][aloc[1]] == 0) return new Game(false, 0);

        for (int[] move : moves) {
            int[] nxtA = new int[]{aloc[0]+move[0], aloc[1]+move[1]};
            if (nxtA[0] < 0 || nxtA[0] >= board.length || nxtA[1] < 0 || nxtA[1] >= board[0].length) continue;
            int nxtV = board[nxtA[0]][nxtA[1]];

            if (nxtV == 1) {
                board[aloc[0]][aloc[1]] = 0;
                Game res = simulation(board, bloc, nxtA);
                board[aloc[0]][aloc[1]] = 1;
                results.add(new Game(!res.win, res.turn+1));
            }
        }

        if (results.size() == 0) return new Game(false, 0);
        else {
            int _min = Integer.MAX_VALUE, _max = 0;
            boolean flag = false;
            for (Game res : results) {
                if (res.win) {
                    flag = true;
                    _min = Math.min(_min, res.turn);
                } else {
                    _max = Math.max(_max, res.turn);
                }
            }
            if (flag) return new Game(true, _min);
            else return new Game(false, _max);
        }
    }   
}
